File: app/src/main/java/com/example/elsafa/Cart/CartFragment.java
package com.example.elsafa.Cart;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.elsafa.HomeActivity;
import com.example.elsafa.R;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.FirebaseFirestore;

import java.util.ArrayList;

import Model.Item;

public class CartFragment extends Fragment implements OnGetCartItem {

    private final FirebaseAuth auth;
    private final FirebaseFirestore fStore;
    private final ArrayList items;
    private CartRecyclerViewAdapter adapter;
    private LinearLayout emptyCart, shopping;
    private View check_out;
    private static TextView totalBillView;
    private double totalBill;
    private CartPresenter presenter;
    private RecyclerView recyclerView;

    public CartFragment() {
        auth = FirebaseAuth.getInstance();
        fStore = FirebaseFirestore.getInstance();
        items = new ArrayList();
        totalBill = 0;
    }

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_cart, container, false);

        recyclerView = root.findViewById(R.id.cart);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setHasFixedSize(true);
        adapter = new CartRecyclerViewAdapter(items, getContext(), this);
        recyclerView.setAdapter(adapter);

        emptyCart = root.findViewById(R.id.empty_cart);
        shopping = root.findViewById(R.id.shopping);
        check_out = root.findViewById(R.id.check_out);
        totalBillView = root.findViewById(R.id.bill_cost);

        presenter = new CartPresenter();
        presenter.setOnGetCartItem(this);
        presenter.getCartItems();

        return root;
    }


    public void updateView() {
        items.clear();
        adapter.notifyDataSetChanged();
        totalBill = 0;
        presenter.getCartItems();
    }


    public void editTotalBill(double price) {
        totalBill += price;
        totalBillView.setText(totalBill + " EGP");
    }

    private void getEmptyCartView() {
        emptyCart.setVisibility(View.VISIBLE);
        shopping.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(getContext(), HomeActivity.class);
                getActivity().startActivity(intent);
                getActivity().finish();
            }
        });
    }

    @Override
    public void onGetCartItemSuccess(Item item) {
        items.add(item);
        adapter.notifyDataSetChanged();
        check_out.setVisibility(View.VISIBLE);
    }

    @Override
    public void onGetCartItemFailed(Exception e) {

    }

    @Override
    public void onCartIsEmpty() {
        check_out.setVisibility(View.INVISIBLE);
        getEmptyCartView();
    }
}

File: app/src/main/java/com/whiteside/cafe/Cart/CartFragment.kt
package com.whiteside.cafe.Cart

import Model.Item
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.whiteside.cafe.HomeActivity
import com.whiteside.cafe.R
import com.whiteside.cafe.Receipt
import com.whiteside.cafe.SignUp.SignUp
import java.util.*

class CartFragment : Fragment(), OnGetCartItem {
    private val auth: FirebaseAuth?
    private val fStore: FirebaseFirestore?
    private val items: ArrayList<*>?
    private var adapter: CartRecyclerViewAdapter? = null
    private var emptyCart: LinearLayout? = null
    private var shopping: LinearLayout? = null
    private var check_out: View? = null
    private var totalBill: Double
    private var presenter: CartPresenter? = null
    private var recyclerView: RecyclerView? = null
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?, savedInstanceState: Bundle?
    ): View? {
        val root = inflater.inflate(R.layout.fragment_cart, container, false)
        recyclerView = root.findViewById(R.id.cart)
        recyclerView.setLayoutManager(LinearLayoutManager(context))
        recyclerView.setHasFixedSize(true)
        adapter = CartRecyclerViewAdapter(items, context, this)
        recyclerView.setAdapter(adapter)
        emptyCart = root.findViewById(R.id.empty_cart)
        shopping = root.findViewById(R.id.cart_shopping)
        check_out = root.findViewById(R.id.check_out)
        totalBillView = root.findViewById(R.id.bill_cost)
        presenter = CartPresenter()
        presenter.setOnGetCartItem(this)
        presenter.getCartItems()
        setCheckOutListener()
        return root
    }

    private fun setCheckOutListener() {
        check_out.setOnClickListener(View.OnClickListener {
            if (auth.getCurrentUser().isAnonymous) {
                val intent = Intent(context, SignUp::class.java)
                startActivity(intent)
            } else {
                val intent = Intent(context, Receipt::class.java)
                intent.putExtra("Total Bill", totalBill)
                startActivity(intent)
            }
        })
    }

    fun updateView() {
        items.clear()
        adapter.notifyDataSetChanged()
        totalBill = 0.0
        presenter.getCartItems()
    }

    fun editTotalBill(price: Double) {
        totalBill += price
        totalBillView.setText("$totalBill EGP")
    }

    private fun getEmptyCartView() {
        emptyCart.setVisibility(View.VISIBLE)
        shopping.setOnClickListener(View.OnClickListener {
            val intent = Intent(context, HomeActivity::class.java)
            activity.startActivity(intent)
            activity.finish()
        })
    }

    override fun onGetCartItemSuccess(item: Item?) {
        items.add(item)
        adapter.notifyDataSetChanged()
        check_out.setVisibility(View.VISIBLE)
    }

    override fun onGetCartItemFailed(e: Exception?) {}
    override fun onCartIsEmpty() {
        check_out.setVisibility(View.INVISIBLE)
        getEmptyCartView()
    }

    companion object {
        private var totalBillView: TextView? = null
    }

    init {
        auth = FirebaseAuth.getInstance()
        fStore = FirebaseFirestore.getInstance()
        items = ArrayList<Any?>()
        totalBill = 0.0
    }
}